### PR TITLE
fix(database): reject wrapped indexes and out-of-range asset values

### LINKED ISSUE #3385: database: reject wrapped indexes and out-of-range asset values
Several Dingo numeric conversions accept values outside their target width and produce wrapped or architecture-dependent state:

- `database/lifecycle/blob_bulk_delete.go`: `start+uint64(batchSize)-1` wraps before the upper bound is applied, so a persisted index near `MaxUint64` selects the wrong deletion range.
- `database/plugin/metadata/sqlstore/operational.go`: signed SQL integers are converted directly to `uint64`, so a negative stored tip becomes a near-maximum chain point.
- `database/block_indexer.go`: the 32-bit `safeIntToUint32` guard compares against an unavailable width and the database package does not compile for 32-bit Go.
- `database/models/asset.go`: `big.Int.Uint64()` silently keeps low bits when an indexed amount exceeds uint64.

Acceptance criteria:

- Check sign, width, and arithmetic overflow before conversion or range construction.
- Return an error without mutating metadata or selecting a wrapped range.
- Cover maximum, negative, and 32-bit build cases.

### PR DESCRIPTION
**Pull request by arepala-uml**
## Summary

Several Dingo numeric conversions accepted values outside their target
width and produced wrapped or architecture-dependent state. This closes
#3385 by checking sign, width, and arithmetic overflow before conversion
or range construction, and by returning an error instead of mutating
metadata or selecting a wrapped range.

## Changes

**`database/lifecycle/blob_bulk_delete.go`**
Extracted the batch-end computation into a new `batchEnd(start, tipID uint64, batchSize int) uint64` helper. The old `start+uint64(batchSize)-1` could wrap past `math.MaxUint64` before `min()` clamped it, so a persisted block index near `MaxUint64` produced an `end` below `start` and corrupted the deletion range. The helper computes `tipID-start` first (always safe, since the loop guarantees `start <= tipID`) and only adds `batchSize` once that's proven not to overflow.

**`database/plugin/metadata/sqlstore/operational.go`**
Added `checkedUint64(value int64) (uint64, error)`, the mirror of the existing `checkedInt64`. Applied it everywhere a signed SQL column was being cast straight to `uint64`: `GetTip` (the case named in the issue — a negative stored slot/block number previously became a near-`MaxUint64` chain point), `GetNetworkState`, `GetBlockNoncesInSlotRange`, `GetLatestBlockNonce`, `GetDatum`, `GetScript`, `GetPParams`, `GetPParamUpdates`, and `epochFromValues` (shared by the `GetEpoch*` family). All of these already returned `error`, so no interface changes were needed; `epochFromValues` gained an `error` return and its 4 callers picked up one more check each.

**`database/block_indexer.go`**
`safeIntToUint32` compared a plain `int` against the untyped constant `math.MaxUint32`, which doesn't fit in a 32-bit `int` — a compile-time "constant overflows int" error on 32-bit platforms. Fixed by comparing via `int64(n)` instead, which is wide enough on every platform while preserving the existing clamp-to-`MaxUint32` behavior.

**`database/plugin/blob/aws/database.go`, `database/plugin/blob/gcs/database.go`**
Not named in the issue, but `writeReverseKey` in both files had the identical `len(key) > math.MaxUint32` pattern and the same 32-bit compile failure. Fixed the same way, since leaving them would mean the module still didn't build on 32-bit.

**`database/models/asset.go`**
Added `checkedUint64FromBigInt(amount *big.Int) (uint64, error)`, guarded by `big.Int.IsUint64()` (false for both negative values and values ≥ 2⁶⁴). `ConvertMultiAssetToModels` now calls this instead of `amount.Uint64()` directly and returns `([]Asset, error)` instead of `[]Asset`.

**`database/models/utxo.go`**
`UtxoLedgerToModel`'s signature changed from `Utxo` to `(Utxo, error)` to propagate `ConvertMultiAssetToModels`'s new error, and it applies the same `checkedUint64FromBigInt` guard to the UTxO's own lovelace amount.

**`database/models/utxo_test.go`, `database/plugin/metadata/sqlstore/transaction_write.go`, `database/plugin/metadata/sqlstore/utxo.go`, `database/transaction.go`**
Mechanical fallout at the 5 call sites of `UtxoLedgerToModel` (`recoverConsumedUtxo`, `SetGenesisTransaction`, `SetTransaction`, `SetGapBlockTransaction`, `AddUtxos`, plus the existing unit test): each already sat inside a function/closure returning `error`, so each picked up `model, err := models.UtxoLedgerToModel(...)` and an `if err != nil` check.

**New test files**
- `database/lifecycle/blob_bulk_delete_internal_test.go` — `batchEnd` at/near `MaxUint64`.
- `database/block_indexer_safe_conversion_test.go` — `safeIntToUint32` at negative, zero, `MaxUint32`, and `MaxUint32+1`.
- `database/models/asset_test.go` — `ConvertMultiAssetToModels`/`UtxoLedgerToModel` at `MaxUint64`, one past it, and negative, plus nil-input cases.
- `database/plugin/metadata/sqlstore/operational_test.go` — `GetTip` against a hand-inserted row with a negative slot/block number, a `MaxInt64` round-trip, and direct `checkedUint64` unit tests.
- `database/plugin/metadata/sqlstore/utxo_test.go` — `AddUtxos` given an overflowing asset amount, asserting it errors and leaves the `utxo`/`asset` tables empty (no partial mutation).

Closes #3385

## Summary by CodeRabbit

- **Bug Fixes**
  - Prevented numeric overflow and truncation when converting asset amounts, UTxO data, and stored metadata values.
  - Invalid negative or oversized values now return clear errors instead of producing corrupted data.
  - Improved deletion batching to safely handle maximum numeric boundaries.
  - Added compatibility fixes for 32-bit platforms.

- **Validation**
  - Expanded regression coverage for overflow handling, boundary values, corrupted metadata, and failed UTxO conversions.

### CODE REVIEW

**Comment by arepala-uml** [comment]
## Test results

### 1. New/changed regression tests (verbose)
All new tests added for this issue, run individually:

| Test | File | Result |
|---|---|---|
| `TestBatchEndDoesNotWrapNearMaxUint64` | `database/lifecycle/blob_bulk_delete_internal_test.go` | ✅ PASS |
| `TestBatchEndClampsToTipIDWhenBatchSizeExceedsRemainder` | same | ✅ PASS |
| `TestBatchEndStopsAtBatchSizeWhenTipIDIsFarther` | same | ✅ PASS |
| `TestBatchEndSingleBlockBatch` | same | ✅ PASS |
| `TestSafeIntToUint32` (5 subtests: negative, zero, typical, `MaxUint32`, `MaxUint32+1`) | `database/block_indexer_safe_conversion_test.go` | ✅ PASS |
| `TestConvertMultiAssetToModelsMaxUint64` | `database/models/asset_test.go` | ✅ PASS |
| `TestConvertMultiAssetToModelsRejectsOverflow` | same | ✅ PASS |
| `TestConvertMultiAssetToModelsRejectsNegative` | same | ✅ PASS |
| `TestConvertMultiAssetToModelsNilMultiAsset` | same | ✅ PASS |
| `TestCheckedUint64FromBigIntRejectsNil` | same | ✅ PASS |
| `TestUtxoLedgerToModelPaymentScript` (4 subtests, pre-existing, updated for new signature) | `database/models/utxo_test.go` | ✅ PASS |
| `TestUtxoLedgerToModelPropagatesAssetOverflowError` | same | ✅ PASS |
| `TestGetTipRejectsNegativeStoredSlot` | `database/plugin/metadata/sqlstore/operational_test.go` | ✅ PASS |
| `TestGetTipRejectsNegativeStoredBlockNumber` | same | ✅ PASS |
| `TestGetTipRoundTripsMaxInt64Slot` | same | ✅ PASS |
| `TestCheckedUint64` | same | ✅ PASS |
| `TestAddUtxosRejectsOverflowAssetWithoutMutation` | `database/plugin/metadata/sqlstore/utxo_test.go` | ✅ PASS |

### 2. Full package suites — `go test -count=1 ./database/... ./ledger/...`
✅ All packages passed:
```
ok  	github.com/blinklabs-io/dingo/database	39.353s
ok  	github.com/blinklabs-io/dingo/database/dbinfo	1.082s
ok  	github.com/blinklabs-io/dingo/database/immutable	0.841s
ok  	github.com/blinklabs-io/dingo/database/lifecycle	20.860s
ok  	github.com/blinklabs-io/dingo/database/models	1.048s
ok  	github.com/blinklabs-io/dingo/database/nodesettings	1.066s
ok  	github.com/blinklabs-io/dingo/database/plugin/blob/badger	2.402s
ok  	github.com/blinklabs-io/dingo/database/plugin/blob/internal/blobbackup	1.162s
ok  	github.com/blinklabs-io/dingo/database/plugin/blob/internal/compensate	3.764s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata	1.416s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/deferred	1.389s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/internal/certutil	1.617s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/internal/drepquery	1.581s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/internal/utxocond	1.609s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/labelcodec	1.353s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/sqlite	8.845s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/sqlstore	2.844s
ok  	github.com/blinklabs-io/dingo/database/plugin/metadata/sqlstore/migrations	4.525s
ok  	github.com/blinklabs-io/dingo/database/types	1.335s
ok  	github.com/blinklabs-io/dingo/ledger	38.152s
ok  	github.com/blinklabs-io/dingo/ledger/eras	2.131s
ok  	github.com/blinklabs-io/dingo/ledger/forging	2.753s
ok  	github.com/blinklabs-io/dingo/ledger/governance	16.384s
ok  	github.com/blinklabs-io/dingo/ledger/hardfork	1.509s
ok  	github.com/blinklabs-io/dingo/ledger/leader	3.073s
ok  	github.com/blinklabs-io/dingo/ledger/leios	6.489s
ok  	github.com/blinklabs-io/dingo/ledger/rewards	1.045s
ok  	github.com/blinklabs-io/dingo/ledger/snapshot	14.734s
```

### 3. Repo-wide suite — `go test -count=1 ./...`
Two pre-existing, unrelated failures — neither touches any file in this PR, and both were confirmed independent of this change:

- `github.com/blinklabs-io/dingo/bin` — `TestEntrypointForwardsSignalsDuringMithrilBootstrap/resumed_bootstrap_SIGINT`: "Mithril bootstrap child did not become ready." Flaky under system load (spawns a real subprocess and waits on a readiness signal). Re-ran in isolation 4 times after the full-suite run (`-count=3` plus one `-v`): passed every time.
- `github.com/blinklabs-io/dingo/internal/docsparity` — `TestDocumentationDiscoveryIgnoresUntrackedFiles`: fails identically on unmodified `main` (verified via `git stash`/clean checkout before this branch existed); caused by stray untracked files under `.claude/workflows/` in this local environment, unrelated to any file in this PR.

Everything else passed.

### 4. 32-bit build verification (the issue's named regression) — `GOOS=linux GOARCH=386 go build ./database/...`
The three files fixed for the 32-bit compile bug (`database/block_indexer.go`, `database/plugin/blob/aws/database.go`, `database/plugin/blob/gcs/database.go`) now compile clean on 32-bit. The only remaining failure is inside the third-party `gouroboros` dependency (`ledger/conway/rules.go:499`, same `math.MaxUint32`-vs-`int` pattern), which is upstream and out of this repo's scope:
```
# github.com/blinklabs-io/gouroboros/ledger/conway
.../gouroboros@v0.197.0/ledger/conway/rules.go:499:12: math.MaxUint32 (untyped int constant 4294967295) overflows int
```

File: database/block_indexer_safe_conversion_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		want uint32
+	}{
+		{name: "negative clamps to zero", n: -1, want: 0},
+		{name: "zero", n: 0, want: 0},
+		{name: "typical block offset", n: 90_000, want: 90_000},
+		{name: "exactly MaxUint32", n: math.MaxUint32, want: math.MaxUint32},
+		{
+			name: "beyond MaxUint32 clamps",
+			n:    math.MaxUint32 + 1,
+			want: math.MaxUint32,
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply database/block_indexer_safe_conversion_test.go]: Fixed in d26e87ea: replaced the MaxUint32/MaxUint32+1 cases with a `strconv.IntSize == 64` guard using a runtime `int64->int` conversion, so the test package compiles on 32-bit again. Verified with an isolated `GOOS=linux GOARCH=386 go test -c` repro of this exact file.

File: database/lifecycle/blob_bulk_delete.go [hunk 2]
```diff
@@ -45,6 +45,22 @@ func blockIndexID(key []byte) (id uint64, ok bool) {
...
+// substituting DefaultBlockDeleteBatchSize for any batchSize <= 0 before
+// calling in.
+func batchEnd(start, tipID uint64, batchSize int) uint64 {
+	size := uint64(batchSize) // #nosec G115 -- batchSize > 0, see doc comment
+	if tipID-start >= size-1 {
+		return start + size - 1
+	}
+	return tipID
+}
+
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply database/lifecycle/blob_bulk_delete.go:63]: Confirmed and fixed in d26e87ea: added an explicit `if end == tipID { break }` right after `blocksDeleted += batchDeleted`, so the loop stops on the final batch instead of relying on `start <= tipID`. Added `TestDeleteBlocksAfterStopsAtMaxUint64TipWithoutWrapping` (a bounded 5s context so a regression fails fast instead of hanging) — I verified it fails against the unpatched loop (deletes a block far below afterID / times out) and passes with the fix.

File: database/plugin/metadata/sqlstore/operational.go [hunk 3]
```diff
@@ -748,11 +791,15 @@ func (s *Store) GetScript(
...
+	createdSlot, err := checkedUint64(row.CreatedSlot.Int64)
+	if err != nil {
+		return nil, fmt.Errorf("get script: %w", err)
+	}
 	return &models.Script{
 		Hash:        row.Hash,
 		Content:     row.Content,
 		ID:          uint(row.ID),
-		CreatedSlot: uint64(row.CreatedSlot.Int64),
+		CreatedSlot: createdSlot,
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply database/plugin/metadata/sqlstore/operational.go:806]: Confirmed and fixed in d26e87ea: added a `checkedUint8` helper (mirroring `checkedUint64`) and used it for `row.Type.Int64` in `GetScript`, rejecting negative and >255 values instead of narrowing directly. Added `TestGetScriptRejectsOutOfRangeStoredType`, `TestGetScriptRejectsNegativeStoredType`, and a direct `TestCheckedUint8` unit test.

File: database/plugin/metadata/sqlstore/operational.go [hunk 4]
```diff
@@ -1095,19 +1162,27 @@ func epochFromValues(
 	epochID, startSlot sql.NullInt64,
 	nonce, evolvingNonce, candidateNonce, lastEpochBlockNonce []byte,
 	eraID, slotLength, lengthInSlots sql.NullInt64,
-) *models.Epoch {
+) (*models.Epoch, error) {
+	epochId, err := checkedUint64(epochID.Int64)
```
Comment (wolf31o2) [inline database/plugin/metadata/sqlstore/operational.go:1170]: Please extend this checked conversion to `eraID`, `slotLength`, and `lengthInSlots` before narrowing them to `uint`, including the target-width upper bound. A focused current-head probe passed `era_id = -1`; `epochFromValues` returned success with `EraId == MaxUint`. The same silent wrap or width truncation remains for all three fields, so corrupted or legacy epoch rows still escape the SQL-boundary hardening added here.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]

File: database/plugin/metadata/sqlstore/operational.go (hunk 4 above)
Comment (arepala-uml) [reply database/plugin/metadata/sqlstore/operational.go:1170]: Good catch, thanks — fixed in 1435111f. Added `checkedUint(value int64) (uint, error)`, bounded against this platform's actual `uint` width via `^uint(0)` rather than a hardcoded 64-bit assumption (so it correctly rejects narrower-than-int64 overflows on a 32-bit build too), and applied it to `eraID`, `slotLength`, and `lengthInSlots` in `epochFromValues`, plus `EraId` in `GetPParams` for consistency. Added `TestGetEpochRejectsNegativeStoredEraID`/`SlotLength`/`LengthInSlots` via `GetEpoch` (which filters only by `epoch_id`, so these fields reach `epochFromValues` regardless of their stored value — reproducing your era_id = -1 probe) and a direct `TestCheckedUint` unit test.

**Review by wolf31o2 — APPROVED** [review]
